sale_delivery_date/models/sale_order_line.py

```diff
@@ -43,129 +43,79 @@ def _prepare_procurement_values_commitment_date(self, res):
+        date_planned = (
             self.order_id.date_order
-            + timedelta(days=self.customer_lead or 0.0)
-            - timedelta(days=self.order_id.company_id.security_lead)
-        )
-        res["date_planned"] = self._get_date_planned_with_cutoff_time(
-            res["date_planned"]
-        )
-        res["date_planned"] = self._get_date_planned_with_warehouse_calendar(
-            res["date_planned"]
+            + +timedelta(days=customer_lead or 0.0)
+            - timedelta(days=security_lead)
         )
-        res["date_deadline"] = self._get_date_deadline_with_delivery_window(
-            res["date_planned"]
-        )
-        res["date_planned"] = self._get_date_planned_from_date_deadline(
-            res["date_deadline"]
+        partner = self.order_id.partner_shipping_id
+        warehouse = self.order_id.warehouse_id
+        calendar = warehouse.calendar_id
+        date_planned = self._apply_cutoff(date_planned, partner, warehouse)
+        date_planned = self._apply_workload(date_planned, workload_days, calendar)
+        date_deadline = self._apply_delivery_window(date_planned, partner, calendar)
+        date_planned = self._deduce_workload_and_security_lead(
+            date_deadline, partner, warehouse, workload_days, security_lead
         )
+        res.update({"date_deadline": date_deadline, "date_planned": date_planned})
         return res
 
-    def _get_date_planned_with_cutoff_time(self, date_planned, keep_same_day=False):
-        """Apply the cut-off time on a planned date
-
-        The cut-off configuration is taken on the partner if set, otherwise
-        on the warehouse.
+    @api.model
+    def _deduce_workload_and_security_lead(
+        self, date, partner, warehouse, days, security_lead
+    ):
+        """Return the 'date_planned' from the 'date_deadline'.
 
-        By default, if the planned date is the same day but after the cut-off,
-        the new planned date is delayed one day later. The argument
-        keep_same_day forces keeping the same day.
+        Once we know the delivery date of the customer, we are able to
+        compute the final scheduled date of the transfer by taking into
+        account the workload, the WH calendar and the cutoff time.
         """
-        cutoff = self.order_id.get_cutoff_time()
-        partner = self.order_id.partner_shipping_id
+        # Remove the security lead from the date_deadline
+        date_planned = date - timedelta(days=security_lead)
+        # Deduce the workload
+        date_planned = self._apply_workload(date_planned, -days, warehouse.calendar_id)
+        # the correct day has already been computed, only change the cut-off time
+        return self._apply_cutoff(date_planned, partner, warehouse, keep_same_day=True)
+
+    @api.model
+    def _apply_cutoff(self, date, partner, warehouse, keep_same_day=False):
+        cutoff = self.order_id.get_cutoff_time(partner, warehouse)
```

Comment:
Same: `self.order_id` with `@api.model`, send the `order` as parameter?

Reply:
No, this should be independant from the `sale.order*` models since, as you will soon see, those methods will be called from `stock.{move,picking}` soon

Reply:
Also, this `get_cutoff_time` is an `api.model` method now.